x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java

```diff
@@ -2939,4 +2946,29 @@ public void testNamedFunctionArgumentWithUnsupportedNamedParameterTypes() {
             );
         }
     }
+
+    public void testValidJoinPattern() {
+        var basePattern = randomIndexPatterns();
+        var joinPattern = randomIndexPattern(not(WILDCARD_PATTERN));
+        var onField = randomIdentifier();
+        var type = randomFrom("", "LOOKUP ");
+
+        var plan = statement("FROM " + basePattern + " | " + type + " JOIN " + joinPattern + " ON " + onField);
+
+        var join = as(plan, LookupJoin.class);
+        assertThat(as(join.left(), UnresolvedRelation.class).table().index(), equalTo(unquoteIndexPattern(basePattern)));
+        assertThat(as(join.right(), UnresolvedRelation.class).table().index(), equalTo(unquoteIndexPattern(joinPattern)));
+
+        var joinType = as(join.config().type(), JoinTypes.UsingJoinType.class);
+        assertThat(joinType.columns(), hasSize(1));
+        assertThat(as(joinType.columns().getFirst(), UnresolvedAttribute.class).name(), equalTo(onField));
+        assertThat(joinType.coreJoin().joinName(), equalTo("LEFT OUTER"));
+    }
+
+    public void testInvalidJoinPatterns() {
+        expectError(
+            "FROM " + randomIndexPatterns() + " | JOIN my-index-pattern* ON " + randomIdentifier(),
+            "invalid index pattern [my-index-pattern*], * is not allowed in LOOKUP JOIN"
+        );
```

Comment:
I think this can be expanded a little.

Cases I think would be nice to have to ensure our validation works well:
- wildcard character in different positions
- using quotes `"` and `"""`

Reply:
I want to say we should also invalidate usage of [date math](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/date-math-index-names.html) here - or, more completely, for now only allow concrete index names.

@luigidellaquila's PR will prevent date math from being usable, but only after field caps calls have already been made.

So, to summarize:
- I think we can make the invalidation at parse time as aggressive as possible by allowing only valid, concrete index names.
- And we should add testing to demonstrate that this works with nice-ish error messages.

But! That will likely blow the scope of the PR, so I think we should add a follow-up PR - similarly to the invalidation of CCQ.

What do you think @idegtiarenko ?

Reply:
>Cases I think would be nice to have to ensure our validation works well:
>* wildcard character in different positions
>* using quotes " and """

Added in https://github.com/elastic/elasticsearch/pull/120494/commits/9e7f1258fe8eb7d00ca9fa2701c3641231750579

Re second commend, happy to address this as a followup.